server/src/main/java/org/opensearch/cluster/routing/OperationRouting.java

```diff
@@ -274,10 +281,25 @@ public GroupShardsIterator<ShardIterator> searchShards(
                 clusterState.metadata().weightedRoutingMetadata()
             );
             if (iterator != null) {
-                set.add(iterator);
+                shardIterators.computeIfAbsent(iterator.shardId().getIndex(), k -> new ArrayList<>()).add(iterator);
+            }
+        }
+        List<ShardIterator> allShardIterators = new ArrayList<>();
+        for (List<ShardIterator> indexIterators : shardIterators.values()) {
+            if (slice != null) {
+                // Filter the returned shards for the given slice
+                CollectionUtil.timSort(indexIterators);
+                for (int i = 0; i < indexIterators.size(); i++) {
+                    if (slice.shardMatches(i, indexIterators.size())) {
```

Comment:
I am wondering if we should be using shard id instead of `i`?

```
if (slice.shardMatches(indexIterators.get(i).shardId().id(), indexIterators.size())) {
```

Reply:
So, the challenge here is that it's **not necessarily** the shard ID. 

If you've specified routing preferences or something like `_shards=2,4,8`, then the shard iterators returned from `computeTargetedShards` will only contain the shards corresponding to the routing preference or the shard filter. The modulo logic for slice matching pretends those are the "full universe" of shards for slicing purposeses.

Essentially, I'm mimicking the shard-level behavior that happens here: https://github.com/opensearch-project/OpenSearch/blob/0d54c1600b16081c15902a095111a9a530767099/server/src/main/java/org/opensearch/search/slice/SliceBuilder.java#L245-L253 (though that code does the **terrible** thing of using the ordinal positions but pretending that they're `shardId`s). 

IMO, the approach I used (where it's always ordinals, which may happen to be the same as shard IDs) is slightly less hacky.

I should add a comment to explain why it needs to be ordinals and not shard IDs, though.

Reply:
Oh I see, it makes sense but `SliceBuilder` should be using something like `shardIndex` than? Cuz now it uses 
`shardId` explicitly:

```
public boolean shardMatches(int shardId, int numShards) {
```